[MRG] Optimize 3D display (continued)

rth opened the pull request:
## PR Summary

Faster 3D projections using numpy arrays instead if built-in Python objects

Continued https://github.com/matplotlib/matplotlib/pull/6085 (squashed and rebased WeatherGod's rebased version from 2017).

## PR Checklist

- [x] Rebase original PR
- [x] Check consistency of the diff with the original PR
- [x] Make CI pass
- [x] Address review comments
  * Kojoley  [ ]
  * WeatherGod  [ ]
  * anntzer [ ]
  * eric-wieser [ ]
- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] ~~New features are documented, with examples if plot related~~ -> *not applicable*
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way 

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

rth commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -229,23 +245,33 @@ def set_segments(self, segments):
+
+            self._segments3d_data = np.empty((n_segments, 4))
+            self._segments3d_data[:, :3] = np.vstack(segments)
+            self._segments3d_data[:, 3] = 1
```
https://github.com/matplotlib/matplotlib/pull/6085#discussion_r128392253 by @eric-wieser 

> This would be marginally better as
> 
> ```py
> self._segments3d_data  = np.empty((n_segments, 4))
> self._segments3d_data[:,:3] = segments
> self._segments3d_data[:,3] = 1
> ```

Done.

> Also, are you sure you mean "quaternion" and not "homogenous coordinate"?

Not sure, but I removed the "quaternion" term as suggested.

rth commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -229,23 +245,33 @@ def set_segments(self, segments):
+
+            # For coveniency, store a view of the array in the original shape
+            self._segments3d = _array_split(self._segments3d_data[:, :3],
+                                            np.cumsum(self._seg_sizes))
```
https://github.com/matplotlib/matplotlib/pull/6085#discussion_r128392376 by @eric-wieser 

> I think this loop is np.split in some form

Definitely, changed.

rth commented on lib/mpl_toolkits/mplot3d/proj3d.py:
```diff
@@ -130,20 +130,19 @@ def ortho_transformation(zfront, zback):
-    return txs, tys, tzs, tis
+    vecw /= vecw[3]
+    # Integrating tis in the numpy array for optimization purposes
+    vecw[3, :] = tis
```
https://github.com/matplotlib/matplotlib/pull/6085#discussion_r128393125 by @eric-wieser 

> Get rid of this colon - it makes the function fail on 1d inputs

I have not changed it  as I think this will work with scalar and 1D arrays, also CI pass. Unless I'm missing something?

eric-wieser replied:
If `vecw.ndim == 1`, then `vec[3,:]` is an error

rth replied:
I understand - fixed.

rth commented on lib/mpl_toolkits/mplot3d/proj3d.py:
```diff
@@ -130,20 +130,19 @@ def ortho_transformation(zfront, zback):
+    vecw /= vecw[3]
+    # Integrating tis in the numpy array for optimization purposes
+    vecw[3, :] = tis
+    return vecw
```
https://github.com/matplotlib/matplotlib/pull/6085#pullrequestreview-51068714 by @eric-wieser 

> Storing coordinates in the first axis of the array is kind of unusual (see how np.linalg and scipy.misc.imread behave).
> 
> Perhaps these should all be xyz[..., 0] instead of xyz[0]?

Well in the current case `xyz[0]` for default C aligned arrays will produce a contiguous array while `xyz[..., 0]` won't unless I'm mistaken, which might have a performance cost (I haven't tested it).  

Is it essentially the difference between `np.meshgrid` with `indexing="xy"` and `indexing="ij"`? 

I have not changed this, if you think this is critical I will.

eric-wieser replied:
You're correct about contiguity, but the choice of internal storage order and external indexing order is orthogonal.

I'll have to come back to this later to decide if it's important

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -72,6 +72,24 @@ def get_dir_vector(zdir):
+
+    if arr_chunks[-1].size == 0:
+        if not remove_empty:
+            # Preserve the 2D dimensionality the last chunk that can be empty
```
missing "of"

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -141,14 +159,9 @@ def set_3d_properties(self, zs=0, zdir='z'):
-        except TypeError:
-            pass
-        self._verts3d = juggle_axes(xs, ys, zs, zdir)
+        zs = np.broadcast_to(zs, len(xs))
```
Maybe just better to do `xs, ys, zs = np.broadcast_arrays(xs, ys, zs)`, which covers more cases

rth replied:
Broadcasting only `zs` is simpler to follow IMO, unless you think there is a case when `xs`, `ys` also needs to be to be broadcasted? if there is a case when `len(xs) ! = len(ys)`, that is not covered by tests / examples? 

In general wouldn't it be preferable to explicitly fail with a shape mismatch if something goes wrong rather than let `np.broadcast_arrays` magically fix the shapes, but then the issue will be rather difficult to debug..

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -746,21 +773,42 @@ def rotate_axes(xs, ys, zs, zdir):
         return xs, ys, zs
 
 
+def _rotate_axes_vec(xyz, zdir):
```
This will be slower than the original `rotate_axes`, not faster, as it involves copying every element. I'm not sure that it makes sense to vectorize over xyz after all.

eric-wieser commented on lib/mpl_toolkits/mplot3d/proj3d.py:
```diff
@@ -130,20 +130,19 @@ def ortho_transformation(zfront, zback):
-    txs, tys, tzs = vecw[0]/w, vecw[1]/w, vecw[2]/w
-    return txs, tys, tzs
+    vecw /= vecw[3]
+    return vecw[0:3]
```
Why the `0:3`?

rth replied:
If I remove this tests fails, because I think `vecw.shape[0]` and `vec.shape[0]` is 4.

Should I change this to validation of `vec` shape before dot product with `M`?

WeatherGod replied:
Yes, the lengths of those arrays is 4, this is expected for a transform operation. This can be rewritten as `return vecw[:3] / vecw[3]` It'll involve 25% fewer division operations.

rth commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -169,12 +182,15 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
-    seg3d = [juggle_axes(x, y, z, zdir) for (x, y, z) in seg]
-    return seg3d
+    for i, ((x, y), code) in enumerate(pathsegs):
+        seg3d[0:2, i] = x, y
```
Thanks for the suggestion @eric-wieser !

I would like to agree on some benchmarks for this PR (will add a separate comment below)  to evaluate your suggestion and also was wondering if something along the lines of,
```py
seg3d[0:2, :] = [[x, y] for ((x, y), code) in pathsegs]
```
(or splitting it in 2 1D lists) would help...

rth replied:
You are right splitting this in two assignment, is faster (and also comparable to a list comprehension), changed
```ipy
In [1]: import numpy as np

In [2]: x = np.zeros((1000, 3))

In [12]: %%timeit
    ...: for idx, val in enumerate(range(1000)):
    ...:     x[:2, idx] = (val, val)
    ...: 
875 µs ± 3.69 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [11]: %%timeit
    ...: for idx, val in enumerate(range(1000)):
    ...:     x[0, idx] = val
    ...:     x[1, idx] = val
    ...: 
358 µs ± 4.37 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [13]: %%timeit
    ...: x[:2].T[:] = [(idx, idx) for idx in range(1000)]
    ...: 
345 µs ± 2.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Though I this change has no impact in the benchmarks I tried.

rth commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -169,12 +182,15 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
     """Convert a path to a 3D segment."""
+    seg3d = np.ones((3, len(path)))
+    # Works both if zs is and array and a scalar
+    seg3d[2, :] *= zs
```
Absolutely, missed that -- this was also the reason for `np.empty` not working. Fixed both.

anntzer commented on lib/matplotlib/path.py:
```diff
@@ -127,7 +127,8 @@ def __init__(self, vertices, codes=None, _interpolation_steps=1,
         if vertices.ndim != 2 or vertices.shape[1] != 2:
             raise ValueError(
-                "'vertices' must be a 2D list or array with shape Nx2")
+                "'vertices' must be a 2D list or array with shape Nx2, "
```
"must be an array-like of shape (N, 2), not {}".format(vertices.shape)

rth commented:
Thanks a lot for the review @eric-wieser and @anntzer!

I responded part of your comments above. What remains is mostly performance suggestions, and I mentioned earlier I was wondering what would be representative benchmarks to double check those.

 - I can use the data from the figure  above (https://github.com/matplotlib/matplotlib/pull/11577#pullrequestreview-137272496) with `plot_trisurf`
 - I imagine the included examples are too small for this PR to affect them?
 - Are there any other benchmark script that I could use for this?

Generally in terms of methodology, is running,
```py
t0 = time()
plt.plot_<some_3d_method>
plt.draw()
print('dt = {.2f}'.forma(time() - t0))
```
a reasonable way to benchmark the rendering time, or are there better approaches? Thank you!

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -445,19 +468,17 @@ def set_3d_properties(self, zs, zdir):
-            ys = []
-        self._offsets3d = juggle_axes(xs, ys, np.atleast_1d(zs), zdir)
+        offsets = np.hstack([offsets,
+                             (np.ones(len(offsets)) * zs)[:, np.newaxis]])
```
why not do a broadcast_to() like you do elsewhere?

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -636,24 +650,21 @@ def do_3d_projection(self, renderer):
-                 in enumerate(zip(xyzlist, cface, cedge))),
-                key=lambda x: x[0], reverse=True)
+            z_argsort = np.argsort(
+                [self._zsortfunc(xyz[:, 2]) for xyz in xyzlist])[::-1]
```
We should double-check if argsorting is the right approach here for this. numpy uses a different algorithm for argsorting than it does for regular sorting, IIRC. Also, argsorting suffers from non-locality of data as the sorting process wraps up, which causes cache-misses.

rth replied:
According to the docs, `np.argsort` and `np.sort` use quicksort while the builtin `sorted` uses timsort. Timsort [is very fast for pre-sorted data](https://stackoverflow.com/a/19587279/1791279) while there have been report of quicksort being faster implementation wise for random data https://github.com/numba/numba/issues/1383#issuecomment-134539513 .

Is there reason to believe that data would be pre-sorted here? or that this is a performance critical part?

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -736,6 +747,22 @@ def juggle_axes(xs, ys, zs, zdir):
+    if zdir == 'x':
+        return xyz[[2, 0, 1]]
+    elif zdir == 'y':
+        return xyz[[0, 2, 1]]
```
I wonder if it is faster to use `np.roll()` here instead?

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -746,21 +773,42 @@ def rotate_axes(xs, ys, zs, zdir):
+    elif zdir == 'y':
+        return xyz[[2, 0, 1]]
+    elif zdir == '-y':
+        return xyz[[1, 2, 0]]
```
Again, maybe `np.roll()` might be more efficient? Dunno.

rth replied:
No longer applicable as the corresponding function was removed.

WeatherGod commented on lib/mpl_toolkits/mplot3d/proj3d.py:
```diff
@@ -130,20 +130,19 @@ def ortho_transformation(zfront, zback):
     if np.any(tis):
         tis = vecw[1] < 1
-    return txs, tys, tzs, tis
+    vecw /= vecw[3]
```
This is not equivalent. The normalization is applied to the first three rows. Here, you do it to all four rows. This problem is only papered over by the subsequent assignment. You can just do `vecw[:3] /= vecw[3]` instead for 25% fewer division operations.

WeatherGod commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -458,7 +465,7 @@ def __init__(self, *args, zs=0, zdir='z', depthshade=True, **kwargs):
         self.set_3d_properties(zs, zdir)
 
-    def set_sort_zpos(self, val):
+    def set_sort_zpos(self,val):
```
no, there should be a space after the comma

rth replied:
Absolutely, there was another `set_sort_zpos` where the space was missing, I added it then decided to revert PEP8 fixes, but looks like I reverted in the wrong function.

rth commented:
Thanks for the review @WeatherGod . I addressed part of the comments, for the rest regarding performance improvements, trying to put some benchmarks together to evaluate the run time. 

> Eh, this is technically an API change. At the very least, note that it returns a numpy array.

So what's should be the situation about backward compatibility? Do I add a `versionchanged`  sphinx directive to the function docstring and mention it in `doc/api/api_changes.rst`? This module currently have mostly public functions, I imagine it would have make sense for e.g. `paths_to_3d_segments` to be private (possibly with a deprecation cycle?) but that would probably be addressed in a separate PR.

rth commented:
Following @eric-wieser 's suggestion (https://github.com/matplotlib/matplotlib/pull/11577#discussion_r202591857) I removed the vectorized versions of `rotate_axes` and `juggle_axes` -- this indeed didn't have much benefit and the current diff is smaller as a result. 

Unless I missed something, I think all review comments should be addressed except for,
  - the discussion in https://github.com/matplotlib/matplotlib/pull/11577#discussion_r202555288 (cc @eric-wieser )
  - the discussion in https://github.com/matplotlib/matplotlib/pull/11577#discussion_r203226671 (cc @WeatherGod )
  - how to deal with API changes. I added sphinx `versionchanged` entries. In general, I think `path_to_3d_segment*` functions should have been private functions instead of public, as they are only used internally, unless I'm mistaken. It might make sense to make them private in a separate PR, in which case the API changes will be addressed, unless a deprecation cycle should be followed. It also depends on the timeline for your 3.0 release if this can make it there, I guess it's more acceptable to have API changes between 2.x and 3.x than between 3.0 and 3.1. Not sure what I should do about this point. 

Also could someone please remove the "work in progress" tag as this PR can be reviewed. Thanks!

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -226,9 +230,12 @@ def paths_to_3d_segments_with_codes(paths, zs=0, zdir='z'):
-    return np.asarray(segments), np.asarray(codes_list)
+        return np.asarray(segments), np.asarray(codes_list)
+    else:
+        return np.empty((3, 0)), np.array([])
```
This should probably have explicit dtypes. What type are `segments` and `codes_list` normally?

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -558,26 +610,17 @@ def set_zsort(self, zsort):
-        self._segis = segis
+        """Optimize points for projection"""
+
+        self._seg_sizes = [len(c) for c in segments3d]
```
May as well cast to array here

rth commented:
Thanks for additional comments @eric-wieser -- added commits to address them.

anntzer commented:
The PR probably won't make it to 3.0(?)  I would suggest just leaving the old functions as they are and decorate them with `@cbook.deprecated("3.1")` (with an API changes notice), and duplicate them to new private functions with the new behavior.

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -168,51 +180,70 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
-    seg3d = [juggle_axes(x, y, z, zdir) for (x, y, z) in seg]
-    return seg3d, codes
+    codes = np.empty(len(path))
+    xs, ys, codes = zip(*((x, y, code) for (x, y), code in pathsegs))
```
To me it looks like the right thing to do would be to add `path.segments_array`, with a structured dtype of `[('x', float), ('y', float), ('code', int)]` or something. That would save you all these zips, and special cases for when the sequence is empty

rth replied:
Absolutely, my intuition is also that `path.iter_segments` might be the bottleneck and that instead of making slice for every vertices group then making a copy with `.flatten()`, using some form of sliding window in numpy (e.g. https://github.com/numpy/numpy/issues/7753) together with reshaping might produce an equivalent result more efficiently (or just returning a view instead of a copy). However, this would require some more in depth refactoring and I feel that this PR already make a significant number of changes with a measurable performance improvements. So I would rather leave additional enhancements to a subsequent PR (along those mentioned in the original PR https://github.com/matplotlib/matplotlib/pull/6085). 

Would that be acceptable for you?

BTW, yes, I need to handle the case of empty sequences, here. Should be fixed now.

rth replied:
Also more generally, I have done some profiling when running `plot_trisurf` on a large dataset (to produce the figure in the beginning of this issue), results can be found in [bench_results_trisurf.txt](https://github.com/matplotlib/matplotlib/files/2238704/bench.txt) and I cannot find anything related to `path_to_3d_segment` there. 

The above change aims to prevent calling `juggle_axes` on every single vertex as it was done before. As you suggested there may be a way to prevent memory copies, but a) I still think that it would make sense to optimize `path.iter_segments` before it b) given the profiling results, I'm not sure it's a performance critical part of the code (but I could have missed something)

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -168,51 +180,70 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
+    codes = np.empty(len(path))
+    xs, ys, codes = zip(*((x, y, code) for (x, y), code in pathsegs))
+    seg3d = juggle_axes(xs, ys, zs, zdir)
+    return np.array(seg3d).T, codes
```
Similarly, this would become just a `.view` by a dtype with the fields at different offsets, possibly eliminating a copy

dopplershift commented:
FYI: Drop in coverage on tests is due to our lovely test suite that includes code that only executes *sometimes*.

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -229,23 +263,33 @@ def set_segments(self, segments):
-            self._segments3d]
-        segments_2d = [np.column_stack([xs, ys]) for xs, ys, zs in xyslist]
+        if len(self._segments3d) == 0:
+            return 1e9
```
Maybe the old fixme should stay here, with a comment explaining its for back-compat. I'd be happier with a change to `float('inf')` here in some future patch.

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -558,26 +613,17 @@ def set_zsort(self, zsort):
+            # Put all segments in a big array
+            _vec = np.vstack(segments3d)
+            # Add a fourth dimension
+            self._vec = np.hstack([_vec, np.ones((n_segments, 1))]).T
```
More efficient as
```
self._vec = np.empty((n_segments, 4))
self._vec[:3,:] = segments3d
self._vec[3,:] = 1
```

which also eliminates the need for the special case when `len(segments3d) == 0`

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -619,9 +665,11 @@ def do_3d_projection(self, renderer):
+
+        xyzlist = _array_split(xys[:, :3],
+                               np.cumsum(self._seg_sizes),
+                               remove_empty=True)
```
Instead of `remove_empty` here, you could just pass `np.cumsum(self._seg_sizes[:-1])`

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -746,21 +791,22 @@ def rotate_axes(xs, ys, zs, zdir):
+    elif len(c) < num:
+        return np.broadcast_to(mcolors.to_rgba_array(c), (num, 4))
+    else:
+        return mcolors.to_rgba_array(c)
```
What shape is this return path?

eric-wieser commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -162,51 +175,73 @@ def line_2d_to_3d(line, zs=0, zdir='z'):
     Convert paths from a collection object to 3D segments with path codes.
+
+    .. versionchanged :: 3.1
+       Return type changed from a list to a numpy.array
```
More accurately: Tuple of lists to tuple of ndarrays

rth commented:
Going back to this PR, I would really like to get this merged. 

Below is a summary of the situations on benchmarks, API changes, and tests.
Feedback would be very appreciated.

## Benchmarks

<details>

```py
import sys
from pathlib import Path
import cProfile
import pstats
import io

import numpy as np
from skimage.measure import marching_cubes_lewiner
from time import time, sleep

from matplotlib.backends.backend_qt5agg import FigureCanvas
from matplotlib.figure import Figure
import matplotlib.pyplot as plt


from PyQt5 import QtWidgets, QtCore

from mpl_toolkits.mplot3d import Axes3D  # noqa

profile = False


class FigureWidget(QtWidgets.QWidget):
    def __init__(self, parent=None, canvas=None):
        super().__init__(parent)
        self.canvas = canvas
        self.layout = QtWidgets.QVBoxLayout()
        if canvas is not None:
            self.layout.addWidget(self.canvas)
        self.setLayout(self.layout)


class ExampleCanvas(FigureCanvas):
    def __init__(self):
        self.figure = Figure()

        FigureCanvas.__init__(self, self.figure)
        self.ax = self.figure.add_subplot(111, projection="3d")

    def plot(self, X=None):

        if X is None:
            self.draw()
            return

        self.ax.clear()

        print(f"# Input dataset shape : {X.shape}")

        t0 = time()
        verts, faces, normals, values = marching_cubes_lewiner(
            X, level=X.max() * 0.5
        )
        dt = time() - t0
        print(f" - marching_cubes_lewiner: {verts.shape[0]} vertices in {dt:.2f}s")

        if profile:
            pr = cProfile.Profile()
            pr.enable()

        t0 = time()
        self.ax.plot_trisurf(
            verts[:, 0], verts[:, 1], faces, verts[:, 2], color="r", antialiased=True
        )
        self.draw()
        plt.pause(0.001)
        dt = time() - t0
        print(f" - plot_trisurf in {dt:.3f}s")

        t0 = time()
        angles = list(range(0, 360, 72))
        for angle in angles:
            self.ax.view_init(30, angle)
            self.draw()
            plt.pause(0.001)
        dt = time() - t0
        print(f" - rotation {dt/len(angles):.2f} s/frame")

        if profile:
            pr.disable()
            s = io.StringIO()
            ps = pstats.Stats(pr, stream=s)
            ps.sort_stats('tottime')
            # ps.print_stats()
            ps.print_callers()
            print(s.getvalue())


if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    canvas = ExampleCanvas()
    widget = FigureWidget(canvas=canvas)
    canvas.plot()
    basedir = Path(__file__).parent

    app = QtWidgets.QApplication(sys.argv)
    canvas = ExampleCanvas()
    widget = FigureWidget(canvas=canvas)
    canvas.plot()
    widget.show()

    class BenchmarkThread(QtCore.QThread):
        def run(self):
            sleep(1)
            X = np.load(basedir / "data_3d.npy")

            X = np.abs(X)
            for step in [4, 2, 1]:
                Xsl = X[::step, ::step, :]
                canvas.plot(Xsl)

    thread = BenchmarkThread()
    thread.start()

    sys.exit(app.exec_())
```
</details>

run on this 3D dataset [data_3d.npy](https://s3-eu-west-1.amazonaws.com/public-sym/data_3d.npy) (130 MB).

The results are below,
**master**
```
# Input dataset shape : (150, 150, 95)
 - marching_cubes_lewiner: 3632 vertices in 0.08s
 - plot_trisurf in 0.751s
 - rotation 0.54 s/frame
# Input dataset shape : (300, 300, 95)
 - marching_cubes_lewiner: 8640 vertices in 0.33s
 - plot_trisurf in 1.729s
 - rotation 1.27 s/frame
# Input dataset shape : (600, 600, 95)
 - marching_cubes_lewiner: 22174 vertices in 1.18s
 - plot_trisurf in 4.276s
 - rotation 3.40 s/frame
```
**This PR**
```
# Input dataset shape : (150, 150, 95)
 - marching_cubes_lewiner: 3632 vertices in 0.09s
 - plot_trisurf in 0.478s
 - rotation 0.28 s/frame
# Input dataset shape : (300, 300, 95)
 - marching_cubes_lewiner: 8640 vertices in 0.34s
 - plot_trisurf in 0.973s
 - rotation 0.55 s/frame
# Input dataset shape : (600, 600, 95)
 - marching_cubes_lewiner: 22174 vertices in 1.15s
 - plot_trisurf in 2.369s
 - rotation 1.24 s/frame
```
so this speeds up `plot_trisurf` initial drawing by 1.5-1.7x and then rendering of the rotation by 2-2.7x for different input types. It is a lot.

## Handle API changes

This PR changes quite a vew functions to return numpy arrays instead of lists.

If I wollow @anntzer 's suggestion (https://github.com/matplotlib/matplotlib/pull/11577#issuecomment-407192061), it would mean deprecating,
 - `art3d.path_to_3d_segment`
 - `art3d.paths_to_3d_segments`
 - `art3d.path_to_3d_segment_with_codes`
 - `art3d.paths_to_3d_segments_with_codes`
 - `art3d.line_collection_2d_to_3d`

then because we cange `proj3d.proj_transform_vec` and `proj3d.proj_transform_vec_clip` we would also have to deprecate (since the output would change),

 - `proj3d.proj_transform`
 - `proj3d.proj_transform_clip`
 - `proj3d.transform`
 - `proj3d.proj_trans_points`
 - `proj3d.proj_trans_clip_points`

I'm not sure what you think the ideal public API of `mplot3d` should look like and whether this would go in the right direction. In any case, if you agree with this, I think these deprecations might better done is a separate PR (prior to merging this one).

## Fix CI

So CI was passing fine a few commits ago, but then after I merged master, I started to get two test failures,
```
E           matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 1.384):
E           	result_images/test_mplot3d/bar3d_shaded.png
E           	result_images/test_mplot3d/bar3d_shaded-expected.png
[...]
E           matplotlib.testing.exceptions.ImageComparisonFailure: images not close (RMS 0.003):
E           	result_images/test_mplot3d/bar3d_notshaded.png
E           	result_images/test_mplot3d/bar3d_notshaded-expected.png
```
For instance,
**bar3d_shaded-expected.png**
![bar3d_shaded-expected](https://user-images.githubusercontent.com/630936/50149469-90069080-02bb-11e9-824f-300272f5a874.png)
**bar3d_shaded.png**
![bar3d_shaded](https://user-images.githubusercontent.com/630936/50149484-9ac12580-02bb-11e9-9e2e-fcbe0d136da2.png)
**bar3d_shaded-failed-diff.png**
![bar3d_shaded-failed-diff](https://user-images.githubusercontent.com/630936/50149490-a280ca00-02bb-11e9-842b-2f6b584f4797.png)

would you have suggestions where this could be coming from?

Apart for that I think I addressed most comments, apart for possible performance improvements in https://github.com/matplotlib/matplotlib/pull/11577/files/ad0e783d17be4dce4d3ee6d84807d388a3bbde14#diff-fc54f86f287bf203dfa9e78102e2678b, https://github.com/matplotlib/matplotlib/pull/11577#pullrequestreview-139734253 by @eric-wieser .

I think any additional performance improvements should be done in a follow up PR. Now I'm mostly concerned about the API deprecations and making CI pass, in order to make this mergable.

WeatherGod commented:
I don't know if we need to completely deprecate all those functions. What if instead of deprecating them, we modify them to output lists if the input was a list (and deprecate that feature as we transition to allow only numpy arrays)? Note, I am not against changing the API, because these benefits are very compelling, and the list-based approach is so outdated. I am just trying to get a feel for our options.

As for the test failure, the image difference is with the bar with no height. Perhaps there is a numerical instability somewhere that is causing the normal vector to flip?

eric-wieser commented:
For the bar with no height, it's possible that you reversed the order of a list of polygons somewhere, but numerical instability would also make sense.

I think in that case it would be fine to change the expected image

anntzer commented on lib/mpl_toolkits/mplot3d/art3d.py:
```diff
@@ -71,6 +71,24 @@ def get_dir_vector(zdir):
+    if arr_chunks[-1].size == 0:
+        if not remove_empty:
+            # Preserve the 2D dimensionality of the last chunk that can be
+            # empty (numpy <=1.10 replaces empty chunks by a 1D empty array)
```
we require numpy>=1.11 now.

anntzer commented:
> I don't know if we need to completely deprecate all those functions. What if instead of deprecating them, we modify them to output lists if the input was a list (and deprecate that feature as we transition to allow only numpy arrays)? Note, I am not against changing the API, because these benefits are very compelling, and the list-based approach is so outdated. I am just trying to get a feel for our options.

Most of these functions are clearly helpers that should never have been part of the public API to start with.  More importantly, adding code to each and every one of them to make them check whether the input is a list and return a list in that case (and then immediately deprecating that) is just code churn; plus it *still* breaks backcompat if someone was passing an array in and expecting a list out (which was the previous behavior), so it doesn't actually help anyone.

rth commented:
Thanks for the feedback. Made a PR to deprecate these helper functions in https://github.com/matplotlib/matplotlib/pull/13030

eric-wieser commented:
Rebase conflicts are likely with #16472

rth commented:
Sorry I don't have the availability to continue working on this. So this PR can be marked as stalled (or a more appropriate label). 

If someone wants to continue it, they should certainly feel free to do so.

tacaswell commented:
Given that this has been stalled for 2 years and we have had a fair amount of refactoring to the mplot3d code base I am going to close this, however it may be worth re-considering the ideas in these changes to improve performance.

Thankyou for your effort on this @rth , @eric-wieser and @Kojoley 

@scottshambaugh
